index.js

```diff
@@ -51,8 +55,21 @@ function propertyIsUnsafe(target, key) {
 			&& Object.propertyIsEnumerable.call(target, key)) // and also unsafe if they're nonenumerable.
 }
 
+// Retrieves either a new object or the appropriate target object to mutate.
+function getDestinationObject(target, source, options) {
+	const targetDefined = typeof target !== 'undefined'
+	const isArray = Array.isArray(target) || Array.isArray(source)
+	const doMerge = options && (options.mergeWithTarget || options.clone === false)
+
+	if (targetDefined && doMerge) {
+		return Array.isArray(target) ? firstArrayEntry(target) : target
+	}
+
+	return isArray ? [] : {};
```

Comment:
The destination object should never be an array.

This is how I would define this function:

```js
function getDestinationObject(target, options) {
	const targetDefined = typeof target !== 'undefined'
	const doMerge = options && (options.mergeWithTarget || options.clone === false)

	if (targetDefined && doMerge) {
		return target
	}

	return {};
}
```

Reply:
For me, the array check was required because during recursion it's possible that there is a property being handled whose value is `undefined`, but we know that the source value is an array, so we need to instantiate an empty object that can be merged into. If I remove this, nested arrays will fail to merge because an array source will try to be merged into an object target.

Reply:
Yeah sorry, I was thrown off by the function `mergeObject`'s name. I assumed that it's parameters were always objects but in actual fact it merges any 2 values (or returns the latter one if they can't be merged).

Reply:
Wait no, that's not right. `mergeObject` does only ever receive non-array objects. `target` and `source` should never be arrays in `mergeObject` - arrays should be handled by `options.arrayMerge`.

Reply:
@creativeux I suspect you'll need to update the logic in `deepmerge` itself to handle the case of `target` being `undefined`.

https://github.com/TehShrike/deepmerge/blob/master/index.js#L87-L93